lib/App/Critique/Command/collect.pm

```diff
@@ -42,14 +42,7 @@ sub execute {
 
     my $filter;
     if ( my $f = $opt->filter ) {
-        if ( ref $f eq 'CODE' ) {
-            $filter = $f;
-        }
-        else {
-            $filter = $opt->invert
-                ? sub { $_[0]->stringify !~ /$f/ }
-                : sub { $_[0]->stringify =~ /$f/ };
-        }
+        $filter = file_filter($opt,'file_filter_regex');
     }
```

Comment:
So, if you read the comments inline below, I think we basically only need one function in the plugin, which would look something like this for producing the equivalent of the `no_violations` call.
```
my $filter = file_filter(
    # extract the Path::Tiny object from the App::Critique::Session::File object (as needed)
    path      => $file->path, 
    # the predicate call back would get the path object as its first argument
    predicate => sub { scalar $session->perl_critic->critique( $_[0] ) }, 
    # success and failure callbacks
    success   => sub { info('Found %d violation(s), keeping file (%s) ', @_) },
    failure   => sub { info('Found no violation, pruning file (%s)', @_) },
);
```
and the `regex` one would look like:
```
my $filter = file_filter(
    # extract the Path::Tiny object from the App::Critique::Session::File object (as needed)
    path      => $file->path, 
    # the predicate call back would get the path object as its first argument
    predicate => sub { 
        my $f = $opt->filter || return 0; 
        return $opt->invert ? $_[0] !~ /$f/ : $_[0] =~ /$f/; 
    }, 
    # success and failure callbacks
    success   => sub { 
        shift; # we do not care about the $is_match variable here, so shift it off
        info('Matched, keeping file (%s)', @_) 
    },
    failure   => sub { info('Not matched, pruning file (%s)', @_) },
)
```
Make sense?